pkg/apis/traceflow/v1/types.go

```diff
@@ -0,0 +1,158 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type DropReason string
+
+type Phase string
+
+const (
+	Initial Phase = "Initial"
+	Running Phase = "Running"
+	Success Phase = "Success"
+	Timeout Phase = "Timeout"
+	Error   Phase = "Error"
+)
+
+type Component string
+
+const (
+	SpoofGuard    Component = "SpoofGuard"
+	LB            Component = "LB"
+	Routing       Component = "Routing"
+	NetworkPolicy Component = "NetworkPolicy"
+	Forwarding    Component = "Forwarding"
+)
+
+type Action string
+
+const (
+	Delivered Action = "Delivered"
+	Received  Action = "Received"
+	Forwarded Action = "Forwarded"
+	Dropped   Action = "Dropped"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type Traceflow struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   `json:"spec"`
+	Status `json:"status"`
+}
+
+// Spec describes the spec of the traceflow.
+type Spec struct {
+	Source      `json:"source"`
+	Destination `json:"destination"`
+	Packet      `json:"packet"`
+}
+
+// Source describes the source spec of the traceflow.
+type Source struct {
+	Namespace string `json:"namespace,omitempty"` // Namespace is the source namespace.
+	Pod       string `json:"pod,omitempty"`       // Pod is the source pod.
+}
+
+// Destination describes the destination spec of the traceflow.
+type Destination struct {
+	Namespace string `json:"namespace,omitempty"` // Namespace is the destination namespace.
+	Pod       string `json:"pod,omitempty"`       // Pod is the destination pod, exclusive with destination service.
+	Service   string `json:"service,omitempty"`   // Service is the destination service, exlusive with destination pod.
+	IP        string `json:"IP,omitempty"`        // IP is the destination IP.
+}
+
+// IPHeader describes spec of an IPv4 header. IPv6 not supported yet.
+type IPHeader struct {
+	SrcIP    string `json:"srcIP,omitempty"`    // SrcIP is the source IP.
+	Protocol int32  `json:"protocol,omitempty"` // Protocol is the IP protocol.
+	TTL      int32  `json:"ttl,omitempty"`      // TTL is the IP TTL.
+	Flags    int32  `json:"flags,omitempty"`    // Flags is the flags for IP.
+}
+
+// TransportHeader describes spec of a TransportHeader.
+type TransportHeader struct {
+	ICMPEchoRequestHeader `json:"icmpEchoRequestHeader"`
+	UDPHeader             `json:"udpHeader"`
+	TCPHeader             `json:"tcpHeader"`
+}
```

Comment:
Why using anonymous fields here? The fields are exclusive with each other and have common sub-fields: SrcPort, DstPort so we are not going to benefit from promoted fields.
I think it should be the following in K8s style:
```
type TransportHeader struct {
	icmp *ICMPEchoRequestHeader 
	udp *UDPHeader           
	tcp *TCPHeader  
}
```
https://github.com/kubernetes/kubernetes/blob/15e95e48964d4e800f7210a249f9122ee7d79fcc/pkg/apis/core/types.go#L58
https://github.com/kubernetes/kubernetes/blob/15e95e48964d4e800f7210a249f9122ee7d79fcc/pkg/apis/core/types.go#L2093
https://github.com/kubernetes/kubernetes/blob/15e95e48964d4e800f7210a249f9122ee7d79fcc/pkg/apis/core/types.go#L1776

Reply:
Updated.